File: domain/ledger/index_manager.py
"""IndexManager: blind index of durations by date and title.

A lightweight key-value cache mapping dates to title-to-duration maps.
Derived from the ledger chain and can be fully rebuilt if lost.
"""

import json
from typing import Dict, Any, Optional

from storage.index_store import AbstractIndexStore


class IndexManager:
    """Manages a blind index of {date: {title: total_duration_ms}}.

    Thread-safe via assumption of single-writer pattern (all mutations
    go through the store).

    The index is purely derived data — it can be fully rebuilt from
    the ledger chain using rebuild_from_chain().

    When *crypto* is provided, the index is encrypted at rest using
    AES-128-CTR. Legacy plaintext indices are auto-detected and
    upgraded on next write.
    """

    def __init__(self, store: AbstractIndexStore, crypto=None):
        self.store = store
        self._crypto = crypto
        self._cache: Dict[str, Any] = {}
        self._load()

    def _load(self):
        """Load index from store into memory cache.

        Detects encrypted vs legacy plaintext format automatically:
        - Dict with ``_enc`` key → decrypt
        - Plain dict → legacy, use as-is
        - Empty/falsy → start with empty cache
        """
        stored = self.store.read_index()
        if not stored:
            self._cache = {}
            return
        if isinstance(stored, dict) and "_enc" in stored:
            # Encrypted format
            if self._crypto is not None:
                try:
                    plain = self._crypto.decrypt(stored["_enc"])
                    self._cache = json.loads(plain)
                except Exception:
                    self._cache = {}
            else:
                self._cache = {}
        elif isinstance(stored, dict):
            # Legacy plaintext format
            self._cache = dict(stored)
        else:
            self._cache = {}

    def _flush(self):
        """Write in-memory cache back to store.

        Encrypts the full index dict as JSON when crypto is available.
        Uses ``{"_enc": "<hex_ciphertext>"}`` wrapper format so
        legacy plaintext readers can skip encrypted blobs cleanly.
        """
        if self._crypto is not None:
            plain = json.dumps(self._cache, sort_keys=True)
            encrypted = self._crypto.encrypt(plain)
            self.store.write_index({"_enc": encrypted})
        else:
            self.store.write_index(dict(self._cache))

    def reload(self):
        """Reload cache from the underlying store.

        Call this when an external component may have written to the
        store directly (e.g., legacy code paths).
        """
        self._cache = {}
        self._load()

    def get_all(self) -> Dict[str, Any]:
        """Return a copy of the full index."""
        return dict(self._cache)

    def update(self, date: str, title: str, duration_delta: int):
        """Add or subtract duration for a title on a given date.

        If duration_delta is negative and causes the total to go to
        zero or below, the title entry is removed from that date's
        dict. If the date dict becomes empty, the date is removed.

        Args:
            date: ISO date string (YYYY-MM-DD).
            title: Activity title.
            duration_delta: Duration in ms to add (positive) or
                subtract (negative).
        """
        if date not in self._cache:
            if duration_delta <= 0:
                return
            self._cache[date] = {}

        old = self._cache[date].get(title, 0)
        new = old + duration_delta

        if new <= 0:
            # Remove the title entry; if date is now empty, remove the date too
            if title in self._cache[date]:
                del self._cache[date][title]
            if not self._cache[date]:
                del self._cache[date]
        else:
            self._cache[date][title] = new

        self._flush()

    def query(self, from_date: str, to_date: str) -> Dict[str, int]:
        """Aggregate durations by title over a date range.

        Args:
            from_date: Start date (inclusive), ISO format.
            to_date: End date (inclusive), ISO format.

        Returns:
            Dict of {title: total_ms} over the range. Empty dict if
            no data or from_date > to_date.
        """
        if from_date > to_date or not self._cache:
            return {}

        result: Dict[str, int] = {}
        for date_str, titles in self._cache.items():
            if from_date <= date_str <= to_date:
                for title, duration in titles.items():
                    result[title] = result.get(title, 0) + duration
        return result

    def clear(self):
        """Clear all index data."""
        self._cache = {}
        self._flush()

File: phpoc-web/src/ledger/index_manager.js
/**
 * IndexManager — blind index of durations by date and title.
 *
 * A lightweight key-value cache mapping dates to title-to-duration maps.
 * Derived from the ledger chain and can be fully rebuilt if lost.
 *
 * Uses StorageBackend key convention "ledger:index".
 *
 * When *crypto* is provided, the index is encrypted at rest using
 * AES-128-CTR. Legacy plaintext indices are auto-detected and
 * upgraded on next write.
 *
 * Usage:
 *   import { IndexManager } from './index_manager.js';
 *   const index = new IndexManager(store, crypto);
 *   index.update('2026-01-15', 'Morning Run', 3600000);
 *   const result = index.query('2026-01-01', '2026-01-31');
 */

const INDEX_KEY = 'ledger:index';

export class IndexManager {
  /**
   * @param {import('../sync/storage.js').StorageBackend} store - StorageBackend instance.
   * @param {object} [crypto] - Optional crypto service with encryptWithCachedKey/decryptWithCachedKey.
   */
  constructor(store, crypto) {
    this.store = store;
    this._crypto = crypto || null;
    /** @type {object} Internal cache: {date: {title: total_duration_ms}} */
    this._cache = {};
  }

  /**
   * Write in-memory cache back to store.
   * Encrypts when crypto is available.
   * @returns {Promise<void>}
   */
  _flush() {
    if (this._crypto) {
      const plain = JSON.stringify(this._cache);
      const encrypted = this._crypto.encryptWithCachedKey(plain);
      return this.store.set(INDEX_KEY, { _enc: encrypted });
    }
    return this.store.set(INDEX_KEY, JSON.parse(JSON.stringify(this._cache)));
  }

  /**
   * Reload cache from the underlying store.
   *
   * Auto-detects encrypted vs legacy plaintext format:
   * - Object with _enc key → decrypt
   * - Plain object → legacy, use as-is
   * - Empty/falsy → start with empty cache
   *
   * @returns {Promise<void>}
   */
  async reload() {
    this._cache = {};
    const stored = await this.store.get(INDEX_KEY);
    if (!stored) return;

    if (typeof stored === 'object' && !Array.isArray(stored) && stored._enc) {
      // Encrypted format
      if (this._crypto) {
        try {
          const plain = this._crypto.decryptWithCachedKey(stored._enc);
          if (plain) {
            this._cache = JSON.parse(plain);
          }
        } catch {
          this._cache = {};
        }
      }
    } else if (typeof stored === 'object' && !Array.isArray(stored)) {
      // Legacy plaintext format
      this._cache = JSON.parse(JSON.stringify(stored));
    }
  }

  /**
   * Return a copy of the full index.
   * @returns {object} Deep copy of {date: {title: total_duration_ms}}.
   */
  getAll() {
    return JSON.parse(JSON.stringify(this._cache));
  }

  /**
   * Add or subtract duration for a title on a given date.
   *
   * If durationDelta is negative and causes the total to go to
   * zero or below, the title entry is removed from that date's
   * dict. If the date dict becomes empty, the date is removed.
   *
   * @param {string} date - ISO date string (YYYY-MM-DD).
   * @param {string} title - Activity title.
   * @param {number} durationDelta - Duration in ms to add (positive) or subtract (negative).
   */
  update(date, title, durationDelta) {
    if (!this._cache[date]) {
      if (durationDelta <= 0) {
        return;
      }
      this._cache[date] = {};
    }

    const old = this._cache[date][title] || 0;
    const newVal = old + durationDelta;

    if (newVal <= 0) {
      // Remove the title entry; if date is now empty, remove the date too
      delete this._cache[date][title];
      if (Object.keys(this._cache[date]).length === 0) {
        delete this._cache[date];
      }
    } else {
      this._cache[date][title] = newVal;
    }

    return this._flush();
  }

  /**
   * Aggregate durations by title over a date range.
   *
   * @param {string} fromDate - Start date (inclusive), ISO format.
   * @param {string} toDate - End date (inclusive), ISO format.
   * @returns {object} Dict of {title: total_ms} over the range. Empty if no data.
   */
  query(fromDate, toDate) {
    if (fromDate > toDate) {
      return {};
    }

    const result = {};
    for (const [dateStr, titles] of Object.entries(this._cache)) {
      if (fromDate <= dateStr && dateStr <= toDate) {
        for (const [title, duration] of Object.entries(titles)) {
          result[title] = (result[title] || 0) + duration;
        }
      }
    }
    return result;
  }

  /**
   * Clear all index data.
   */
  clear() {
    this._cache = {};
    return this._flush();
  }
}
